### PR TITLE
Logic Challenge Revamp

### PR DESCRIPTION
**Pull request by s-ruby**
- Re-order exercises so compile time (syntax) errors appear on top.
- Add assert statements and comments to clarify exactly how logical bugs should be solved. 
- Added variable value to print statements for readability 
- The program will stop at each exercise (snippet) until the bug is fixed to help students work through them sequentially and avoid terminal pollution. _i.e only completed snippets and the snippet in progress will print._
- Updated `logic_challenge.md` with explanation about syntax and assertion errors
- Updated solutions file to reflect changes to challenge

**Questions:**
1. Is the bug too explicit with the assertion? i.e is the exercise still challenging?
2. Do we want to introduce `assert()` this early in the course?

### CODE REVIEW

File: challenges/06_bootcamp_logic/logic_challenge_3.md [hunk 1]
```diff
@@ -12,7 +12,14 @@ In this challenge, we are providing you 2 incomplete scripts - `snippets_challen
 
 ### Overview of the challenge
 
-There are 5 code snippets in `snippets_challenge.py` here that throw errors or give incorrect outputs as of now. Based on the lecture, try to debug these code snippets so that the errors are resolved and we have the correct output.
+There are 5 code snippets in `snippets_challenge.py` here that throw errors or give incorrect outputs as of now. Based on the lecture, try to debug these code snippets so that the errors are resolved and we have the correct output. 
+
+Pay attention to the types of errors displayed in the terminal. 
+* Syntax errors means the program cannot compile.
```
Comment (pab2163) [inline challenges/06_bootcamp_logic/logic_challenge_3.md:18]: I'm not sure if mentioning compiling will be helpful or would just confuse students more. My instinct might be to just leave this out for now?

File: challenges/06_bootcamp_logic/logic_challenge_3.md [hunk 2]
```diff
@@ -12,7 +12,14 @@ In this challenge, we are providing you 2 incomplete scripts - `snippets_challen
...
+* `AssertionError` means there is a logical error in the code caused by `assert()`. 
+* `assert()` is a way to test if a condition in your code returns `True`.
```
Comment (pab2163) [inline challenges/06_bootcamp_logic/logic_challenge_3.md:20]: @ashiq-r31 I could go either way on the assertions, what do you think? If we have assertions in the challenge, my instinct would be to mention them in the lesson as well.

File: challenges/06_bootcamp_logic/snippets_challenge.py [hunk 3]
```diff
@@ -4,60 +4,72 @@
...
-elif z > x and z < y
-    print("z is between x and y")
+m = "GOAT"
+n = "goat"
 
-else
-    print("z is greater than y")
+o = (m == n)
 
+print (f"The value of o ({o}) is True since Python is case-insensitive.")
```
Comment (pab2163) [inline challenges/06_bootcamp_logic/snippets_challenge.py:71]: I don't want to accidentally confuse students into thinking python actually IS case-insensitive here. Is there a way we can better scaffold thinking about case sensitivity?

File: challenges/06_bootcamp_logic/snippets_challenge.py [hunk 4]
```diff
@@ -4,60 +4,72 @@
...
-u = 5
-v = 2
+#modify exactly one boolean operator in the assignment of d, so that d evaluates to False
+#replace the print statement with your own explanation
 
-if u * v = 10:
-    print("The product of a and b is 10")
-else:
-    print("The product of a and b is not 10")
+d = ((5 > 7) or not (8 < 20) or True)
```
Comment (pab2163) [inline challenges/06_bootcamp_logic/snippets_challenge.py:54]: this example with two separate `or` statements might trip students up. maybe we could simplify this to just one

File: challenges/06_bootcamp_logic/snippets_challenge.py [hunk 5]
```diff
@@ -4,60 +4,72 @@
...
+print("TO DO: Explain how d is set to False")
```
Comment (pab2163) [inline challenges/06_bootcamp_logic/snippets_challenge.py:55]:
```suggestion
# TO DO: Explain how d is set to False
```

**Review by pab2163 — COMMENTED** [review]
This looks great! Definitely much clearer than before, and nice job with the design so if done in order, the cmd line printout should be informative to the problem students are working on.

Notes:
- I'm still a little unsure of whether incorporating assertions here is a good idea -- let's check in about this and make a decision. 
- Can you make it clear in both the markdown/challenge script that students should make sure they correct each snippet in order? I think most will do this anyway, but just in case. 
-See below for a few line edits

**Comment by pab2163** [comment]
Sorry, I didn't answer question 1. Since this is one of the first times students will be responsible for debugging code someone else wrote, I think it will be challenging enough, even if the bugs seem very 'easy'/explicit

File: challenges/06_bootcamp_logic/logic_challenge_3.md (hunk 1 above)
Comment (ashiq-r31) [reply challenges/06_bootcamp_logic/logic_challenge_3.md:18]: Yeah simplistic language like "cannot run" might be better here

File: challenges/06_bootcamp_logic/logic_challenge_3.md (hunk 2 above)
Comment (ashiq-r31) [reply challenges/06_bootcamp_logic/logic_challenge_3.md:20]: @pab2163 I think we could include `assert` in the lecture. The concept is simple enough and we can demonstrate what assert does and leave it at that.

File: challenges/06_bootcamp_logic/snippets_challenge.py (hunk 4 above)
Comment (ashiq-r31) [reply challenges/06_bootcamp_logic/snippets_challenge.py:54]: @pab2163 agreed

File: challenges/06_bootcamp_logic/snippets_challenge.py (hunk 3 above)
Comment (ashiq-r31) [reply challenges/06_bootcamp_logic/snippets_challenge.py:71]: yeah I'd leave this as "case-sensitive"

**Review by ashiq-r31 — CHANGES_REQUESTED** [review]
Awesome stuff. Adding `assert` was a nice touch.

**Comment by pab2163** [comment]
Great! Just following up -- the plan will be to keep the assertions, and I will update the lecture right before this with a demonstration for what `assert()` does

File: challenges/06_bootcamp_logic/snippets_challenge.py (hunk 4 above)
Comment (s-ruby) [reply challenges/06_bootcamp_logic/snippets_challenge.py]: done!

File: challenges/06_bootcamp_logic/logic_challenge_3.md (hunk 1 above)
Comment (s-ruby) [reply challenges/06_bootcamp_logic/logic_challenge_3.md]: done!

**Review by pab2163 — APPROVED** [review]
thanks, looks great to me!

**Review by ashiq-r31 — APPROVED** [review]
👍 good stuff

**Pull request merged**